tiledb/api/c_api/vfs/vfs_api_external.h

```diff
@@ -607,6 +607,48 @@ TILEDB_EXPORT capi_return_t tiledb_vfs_ls(
     int32_t (*callback)(const char*, void*),
     void* data) TILEDB_NOEXCEPT;
 
+/**
+ * Visits the children of `path` recursively, listing all nested objects in
+ * a single pass. Offsets are in Arrow format [r1, r2, ..., rN] where r1 is the
+ * beginning of the first result and r2 is the beginning of the second. The
+ * final offset rN marks the end of the last result. The length of result N can
+ * be retrieved with rN - rN-1.*
+ * **Example:**
+ *
+ * @code{.c}
+ * std::string data;
+ * std::vector<uint64_t> offsets;
+ * // Retrieve at most 100 results from 'my_dir'.
+ * tiledb_vfs_ls_recursive(ctx, vfs, "my_dir", callback, &data, &offsets, 100);
+ * for (size_t i = 1; i < offsets.size(); i++) {
+ *   std::string path(data, offsets[i - 1], offsets[i] - offsets[i - 1]);
+ *   printf("%s\n", path.c_str());
+ * }
+ * @endcode
+ *
+ * @param[in] ctx The TileDB context.
+ * @param[in] vfs The virtual filesystem object.
+ * @param[in] path The path in which the traversal will occur.
+ * @param[in,out] callback
+ * The callback function to be applied on every visited object.
+ *     The callback should return `0` if the iteration must stop, and `1`
+ *     if the iteration must continue. It takes as input the currently visited
+ *     path, and user provided buffers for paths and their offsets. The callback
+ *     returns `-1` upon error. Note that `path` in the callback will be an
+ *     **absolute** path.
+ * @param[out] data Data buffer results wrote into by recursive ls.
+ * @param[out] data_off Offset data retrieved from recursive ls.
+ * @return `TILEDB_OK` for success and `TILEDB_ERR` for error.
+ */
+TILEDB_EXPORT capi_return_t tiledb_vfs_ls_recursive(
+    tiledb_ctx_t* ctx,
+    tiledb_vfs_t* vfs,
+    const char* path,
+    int32_t (*callback)(const char*, void*, void*),
+    void* data,
+    void* data_off,
+    int64_t max_count) TILEDB_NOEXCEPT;
```

Comment:
I see what you're saying with the second `void*`. Is this still a concern with the change to a single `void*` as a struct? It seems like after this change the caller would handle the overflow in the callback, like I did here:

https://github.com/TileDB-Inc/TileDB/pull/4266/commits/07bd3d7e78c38d57dfc64f8e2deef3cb07bdfadd#diff-63143b90635ecffd1a8c204f366f452e9a75795f4c405846f4195c371448a3adR997

Otherwise internally we're using `std::string` in our `ls_recursive_cb`, so that would reallocate as we append data